File: tests/zig/tree.zig
const std = @import("std");
const profile = @import("profile.zig");

// a treap keyed on x with random priorities y, split/merge as tree.das and the C++ module do
const Node = struct {
    x: i32,
    y: i32,
    left: ?*Node = null,
    right: ?*Node = null,
};

var gpa: std.mem.Allocator = undefined;
var seed: [4]i32 = .{ 1, 1, 1, 1 };
var result: i32 = 0;

// daslib/random's random_big_int: a 4-lane LCG, two 15-bit halves joined
fn randomBigInt() i32 {
    inline for (0..4) |i| seed[i] = 214013 *% seed[i] +% 2531011;
    const lo = (seed[0] >> 16) & 0x7fff;
    const hi = (seed[1] >> 16) & 0x7fff;
    return lo + (hi << 15);
}

fn merge(lower: ?*Node, greater: ?*Node, out: *?*Node) void {
    if (lower == null) {
        out.* = greater;
    } else if (greater == null) {
        out.* = lower;
    } else if (lower.?.y < greater.?.y) {
        out.* = lower;
        merge(lower.?.right, greater, &lower.?.right);
    } else {
        out.* = greater;
        merge(lower, greater.?.left, &greater.?.left);
    }
}

fn split(orig: ?*Node, lower: *?*Node, greaterOrEqual: *?*Node, value: i32) void {
    if (orig == null) {
        lower.* = null;
        greaterOrEqual.* = null;
    } else if (orig.?.x < value) {
        lower.* = orig;
        split(orig.?.right, &orig.?.right, greaterOrEqual, value);
    } else {
        greaterOrEqual.* = orig;
        split(orig.?.left, lower, &orig.?.left, value);
    }
}

fn hasValue(root: *?*Node, x: i32) bool {
    var lower: ?*Node = null;
    var equal: ?*Node = null;
    var greater: ?*Node = null;
    var equalOrGreater: ?*Node = null;
    split(root.*, &lower, &equalOrGreater, x);
    split(equalOrGreater, &equal, &greater, x + 1);
    const res = equal != null;
    var t: ?*Node = null;
    merge(lower, equal, &t);
    merge(t, greater, root);
    return res;
}

fn insert(root: *?*Node, x: i32) void {
    var lower: ?*Node = null;
    var equal: ?*Node = null;
    var greater: ?*Node = null;
    var equalOrGreater: ?*Node = null;
    split(root.*, &lower, &equalOrGreater, x);
    split(equalOrGreater, &equal, &greater, x + 1);
    if (equal == null) {
        const node = gpa.create(Node) catch profile.fail("tree: out of memory");
        node.* = .{ .x = x, .y = randomBigInt() };
        equal = node;
    }
    var t: ?*Node = null;
    merge(lower, equal, &t);
    merge(t, greater, root);
}

fn erase(root: *?*Node, x: i32) void {
    var lower: ?*Node = null;
    var equal: ?*Node = null;
    var greater: ?*Node = null;
    var equalOrGreater: ?*Node = null;
    split(root.*, &lower, &equalOrGreater, x);
    split(equalOrGreater, &equal, &greater, x + 1);
    merge(lower, greater, root);
    if (equal) |node| gpa.destroy(node);
}

fn freeTree(node: ?*Node) void {
    if (node) |nd| {
        freeTree(nd.left);
        freeTree(nd.right);
        gpa.destroy(nd);
    }
}

noinline fn mainTreeNode() i32 {
    var tree: ?*Node = null;
    var cur: i32 = 5;
    var res: i32 = 0;
    var i: i32 = 1;
    while (i < 20000) : (i += 1) {
        const mode = @rem(i, 3);
        cur = @rem(cur * 57 + 43, 10007);
        if (mode == 0) {
            insert(&tree, cur);
        } else if (mode == 1) {
            erase(&tree, cur);
        } else if (hasValue(&tree, cur)) {
            res += 1;
        }
    }
    freeTree(tree);
    return res;
}

fn run(_: void) void {
    result = mainTreeNode();
}

pub fn main(init: std.process.Init) !void {
    gpa = init.gpa;
    try profile.profile(init, "tree", run, {});
}

File: tests/cpp/tree.cpp
#include "profile.h"
#include <cstdlib>

// a treap keyed on x with random priorities y, split/merge as tree.das does
class Tree {
public:
    Tree() = default;
    ~Tree() { delete mRoot; }
    bool hasValue(int x);
    void insert(int x);
    void erase(int x);
private:
    struct Node {
        Node(int x) : x(x) {}
        ~Node() { delete left; delete right; }
        int x = 0;
        int y = rand();
        Node * left = nullptr;
        Node * right = nullptr;
    };
    using NodePtr = Node *;
    static void merge(NodePtr lower, NodePtr greater, NodePtr & merged);
    static void merge(NodePtr lower, NodePtr equal, NodePtr greater, NodePtr & merged);
    static void split(NodePtr orig, NodePtr & lower, NodePtr & greaterOrEqual, int val);
    static void split(NodePtr orig, NodePtr & lower, NodePtr & equal, NodePtr & greater, int val);
    NodePtr mRoot = nullptr;
};

bool Tree::hasValue(int x) {
    NodePtr lower, equal, greater;
    split(mRoot, lower, equal, greater, x);
    bool res = equal != nullptr;
    merge(lower, equal, greater, mRoot);
    return res;
}

void Tree::insert(int x) {
    NodePtr lower, equal, greater;
    split(mRoot, lower, equal, greater, x);
    if (!equal) equal = new Node(x);
    merge(lower, equal, greater, mRoot);
}

void Tree::erase(int x) {
    NodePtr lower, equal, greater;
    split(mRoot, lower, equal, greater, x);
    merge(lower, greater, mRoot);
    delete equal;
}

void Tree::merge(NodePtr lower, NodePtr greater, NodePtr & merged) {
    if (!lower) { merged = greater; return; }
    if (!greater) { merged = lower; return; }
    if (lower->y < greater->y) {
        merged = lower;
        merge(merged->right, greater, merged->right);
    } else {
        merged = greater;
        merge(lower, merged->left, merged->left);
    }
}

void Tree::merge(NodePtr lower, NodePtr equal, NodePtr greater, NodePtr & merged) {
    merge(lower, equal, merged);
    merge(merged, greater, merged);
}

void Tree::split(NodePtr orig, NodePtr & lower, NodePtr & greaterOrEqual, int val) {
    if (!orig) { lower = greaterOrEqual = nullptr; return; }
    if (orig->x < val) {
        lower = orig;
        split(lower->right, lower->right, greaterOrEqual, val);
    } else {
        greaterOrEqual = orig;
        split(greaterOrEqual->left, lower, greaterOrEqual->left, val);
    }
}

void Tree::split(NodePtr orig, NodePtr & lower, NodePtr & equal, NodePtr & greater, int val) {
    NodePtr equalOrGreater;
    split(orig, lower, equalOrGreater, val);
    split(equalOrGreater, equal, greater, val + 1);
}

static volatile int steps = 20000;

PROFILE_NOINLINE int testTree(int count) {
    Tree tree;
    int cur = 5;
    int res = 0;
    for (int i = 1; i < count; i++) {
        int mode = i % 3;
        cur = (cur * 57 + 43) % 10007;
        if (mode == 0) tree.insert(cur);
        else if (mode == 1) tree.erase(cur);
        else res += tree.hasValue(cur);
    }
    return res;
}

int main() {
    int result = 0;
    profile::run("tree", [&] { result = testTree(steps); profile::keep(result); });
    return result >= 0 ? 0 : 1;
}
